database/src/auth.rs

```diff
@@ -0,0 +1,152 @@
+use aes_gcm::{
+    aead::{generic_array::GenericArray, Aead},
+    AeadInPlace, Aes256Gcm, NewAead,
+};
+use once_cell::sync::OnceCell;
+use rand::Rng;
+use rand::RngCore;
+use std::convert::TryInto;
+
+use warp::http::header::AUTHORIZATION;
+use warp::reject;
+use warp::Filter;
+use warp::Rejection;
+
+use crate::user::User;
+use crate::user::UserID;
+use crate::DbConnection;
+
+#[cfg(all(not(debug_assertions), feature = "null_auth"))]
+std::compile_error!("Cannot disable authentication for non-devel environments.");
+
+pub(crate) const NONCE_LEN: usize = 12;
+pub(crate) const TAG_LEN: usize = 16;
+
+/// This is the secret key with which we sign the cookies.
+// TODO: Generate this at first run to ensure security
+static KEY: OnceCell<[u8; 32]> = OnceCell::new();
+
+pub fn generate_key() -> [u8; 32] {
+    rand::thread_rng().gen()
+}
+
+pub fn set_key(k: [u8; 32]) {
+    KEY.set(k).expect("Failed to set secret_key")
+}
+
+fn get_key() -> &'static [u8; 32] {
+    KEY.get().expect("key must be initialized")
+}
+
+#[derive(Debug)]
+pub struct Wrapper(pub User);
+
+impl Wrapper {
+    pub fn get_user(&self) -> UserID {
+        self.0.id
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum AuthError {
+    Missing,
+    Invalid,
+    InvalidKey,
+    BadCount,
+    DBError,
+    DBQueryError,
+    BadBase64,
+    ShortData,
+    DecryptError,
+    CookieError,
+}
+
+impl warp::reject::Reject for AuthError {}
+
+/// Function encrypts a UserID with a nonce and returns it as a base64 string to be used as a cookie/token.
+pub fn user_cookie_generate(user: UserID) -> String {
```

Comment:
Some tests for the cookie functions would be nice.

Reply:
Done